backend_jobs/ingestion_pipeline/pipeline_lib/database_functions.py

```diff
@@ -0,0 +1,145 @@
+"""
+  Copyright 2020 Google LLC
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+"""
+
+import random
+from datetime import datetime
+import apache_beam as beam
+import firebase_admin
+from firebase_admin import credentials
+from firebase_admin import firestore
+from google.cloud import firestore as cloud_firestore
+import geohash2
+
+IMAGES_COLLECTION=u'test2'
+IMAGES_SUB_COLLECTION=u'pipelineRun'
+
+# pylint: disable=protected-access,attribute-defined-outside-init,arguments-differ,abstract-method
+def initialize_database():
+    """
+    Initializes the project's database for writing/reading/updating/deleting purposes.
+    """
+    if not firebase_admin._apps:
+        firebase_admin.initialize_app(credentials.ApplicationDefault(), {
+        'projectId': 'step-project-ellispis',
+        })
+    return firestore.client()
+
+class UploadToDatabase(beam.DoFn):
+    """
+    Uploads in asynchronously the information that was extracted from each
+    image to the project's database.
+    """
+    def setup(self):
+        self.database_firebase = initialize_database()
+
+    def process(self, element, provider, job_name):
+        """
+        Adds/Updates the project's database to contain documents with image attributes.
+        In addition it adds for each image a sub collection with information
+        on the pipelinerun and the provider.
+        Args:
+            element: ImageAttribute field
+            provider: the provider we are runing now
+            job_name: the job name that we are runing now
+        """
+        doc_ref = self.database_firebase.collection(IMAGES_COLLECTION).document(element.id)
+        doc = doc_ref.get()
+        sub_collection_ref= doc_ref.collection(IMAGES_SUB_COLLECTION)
+        sub_collection_doc_ref= sub_collection_ref.document()
+        if doc.exists:
+            #doc found- image has bean ingested already
+            ingested_runs = doc.to_dict()[u'ingestedRuns']
+            ingested_runs.append(job_name)
+            ingested_providers = doc.to_dict()[u'ingestedProviders']
+            if provider.provider_id not in ingested_providers:
+                ingested_providers.append(provider.provider_id)
+            doc_ref.update({
+               u'ingestedRuns':ingested_runs,
+               u'ingestedProviders': ingested_providers,
+               u'visibility': get_max_visibility(provider.visibility,sub_collection_ref),
```

Comment:
wouldn't it be possible to check only the maximum between the current visibility and the current provider?

Reply:
true, fixed it